Framework/Beamline/inc/MantidBeamline/ComponentInfo.h

```diff
@@ -39,19 +41,51 @@ class MANTID_BEAMLINE_DLL ComponentInfo {
 
 private:
   boost::shared_ptr<const std::vector<size_t>> m_assemblySortedDetectorIndices;
-  boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>> m_ranges;
-  size_t m_size = 0;
+  boost::shared_ptr<const std::vector<size_t>> m_assemblySortedComponentIndices;
+  /// Ranges of component ids that are contiguous blocks of detectors.
+  boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+      m_detectorRanges;
+  /// Ranges of component ids that are contiguous blocks of components.
+  boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+      m_componentRanges;
+  Mantid::Kernel::cow_ptr<std::vector<Eigen::Vector3d>> m_positions;
+  Mantid::Kernel::cow_ptr<std::vector<Eigen::Quaterniond>> m_rotations;
+  const size_t m_size = 0;
+  boost::shared_ptr<DetectorInfo> m_detectorInfo;
 
 public:
   ComponentInfo();
-  ComponentInfo(
-      boost::shared_ptr<const std::vector<size_t>>
-          assemblySortedDetectorIndices,
-      boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>> ranges);
+  ComponentInfo(boost::shared_ptr<const std::vector<size_t>>
+                    assemblySortedDetectorIndices,
+                boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+                    detectorRanges,
+                boost::shared_ptr<const std::vector<size_t>>
+                    assemblySortedComponentIndices,
+                boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+                    componentRanges,
+                boost::shared_ptr<std::vector<Eigen::Vector3d>> positions,
+                boost::shared_ptr<std::vector<Eigen::Quaterniond>> rotations,
+                boost::shared_ptr<DetectorInfo> detectorInfo);
+
   std::vector<size_t> detectorIndices(const size_t componentIndex) const;
+  std::vector<size_t> componentIndices(const size_t componentIndex) const;
```

Comment:
Interesting. I did consider `subTree` in the name, but it also includes the original `componentIndex` argument in the full returns. This is because the client-code often needs to loop over everything including the current `componentIndex`.  Maybe `subTree` is still a better alternative to what I have?